File: Python/0344. Reverse String.py
# Write a function that reverses a string. The input string is given as an array of characters s.
# You must do this by modifying the input array in-place with O(1) extra memory.
#
# Example 1:
#
# Input: s = ["h","e","l","l","o"]
# Output: ["o","l","l","e","h"]
#
# Example 2:
#
# Input: s = ["H","a","n","n","a","h"]
# Output: ["h","a","n","n","a","H"]
#
# Constraints:
#
# 1 <= s.length <= 10^5
# s[i] is a printable ascii character.


# 1)
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        s.reverse()


# 2)
class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        s[:] = s[::-1]

File: JavaScript/0344. Reverse String.js
// Write a function that takes a string as input and returns the string reversed.
//
// Example:
// Given s = "hello", return "olleh".

/**
 * @param {string} s
 * @return {string}
 */

/** 1) Cheating */
const reverseString1 = (s) => {
  return s.split('').reverse().join('');
};

/** 2) Two pointers */
const reverseString = (s) => {
  const l = s.length - 1;
  for (let i = 0; i < ~~(s.length / 2); i++) {
    [s[i], s[l - i]] = [s[l - i], s[i]];
  }
};
